donut/modules/voting/helpers.py

```diff
@@ -0,0 +1,395 @@
+from collections import Counter
+from datetime import datetime
+from itertools import chain, groupby
+import json
+import flask
+from donut.auth_utils import get_user_id
+from donut.misc_utils import generate_random_string
+from donut.modules.groups.helpers import get_group_list_data, is_user_in_group
+from donut.validation_utils import (validate_date, validate_exists,
+                                    validate_in, validate_int)
+from .ranked_pairs import winners
+
+ACCESS_KEY_LENGTH = 64
```

Comment:
The access key is to keep surveys private, so you can create a survey that only people with the link (which is `/<access_key>/take`) can fill out. I didn't want it to be brute-forceable but I guess 32 would probably suffice.

Reply:
yeah, i also mention in the overall comments that you should be verifying viewing/taking/submitting the survey based on user_id anyways, which i think you do? in which case it doesn't matter if they have the link or not, they wouldn't be able to fill it out.

Reply:
Yes, you can also restrict by group which checks the `user_id`. I was mostly emulating Google Forms' functionality, not sure whether it would be useful to send a link out to a survey to a group of people that doesn't exist in the database.